# Overview
This nb is meant to understand how the data looks like, and identify explanatory variables. <br>
From the source, in total there are 21 zones. Primarily there are 20 zones with <i>Zone 21</i> being the sum of all other 20 Zones.<br>
All these Zones belong in the USA.
<br><br>
[Official Data Description](https://www.sciencedirect.com/science/article/pii/S0169207013000745?via%3Dihub)

## Load History File

Load is measured in kW for every hour. <br>
<b>Note:</b> The file contains missing values, as highlighted towards the end of this section.

In [1]:
# Importing the load_history file - supposed to be the target variable

import pandas as pd
load_hist = pd.read_csv(r"C:\Users\singh\Desktop\TUD (All Semesters)\Courses - Semester 6 (TU Dresden)\Thesis Work\Dataset\GEFCom2012\GEFCOM2012_Data\Load\Load_history.csv")
load_hist.head()

zone_id  year  month  day      h1      h2      h3      h4      h5      h6  \
0        1  2004      1    1  16,853  16,450  16,517  16,873  17,064  17,727   
1        1  2004      1    2  14,155  14,038  14,019  14,489  14,920  16,072   
2        1  2004      1    3  14,439  14,272  14,109  14,081  14,775  15,491   
3        1  2004      1    4  11,273  10,415   9,943   9,859   9,881  10,248   
4        1  2004      1    5  10,750  10,321  10,107  10,065  10,419  12,101   

   ...     h15     h16     h17     h18     h19     h20     h21     h22  \
0  ...  13,518  13,138  14,130  16,809  18,150  18,235  17,925  16,904   
1  ...  16,127  15,448  15,839  17,727  18,895  18,650  18,443  17,580   
2  ...  13,507  13,414  13,826  15,825  16,996  16,394  15,406  14,278   
3  ...  14,207  13,614  14,162  16,237  17,430  17,218  16,633  15,238   
4  ...  13,845  14,350  15,501  17,307  18,786  19,089  19,192  18,416   

      h23     h24  
0  16,162  14,750  
1  16,467  15,258  
2  13,315  12,424  
3  13,580  11,727  
4  17,006  16,018  

[5 rows x 28 columns]

In [2]:
load_hist.shape

(33000, 28)

The "h#" stands for the hour number. There are 4 explanatory variables <zone_id, year, month, day> in addition.

In [3]:
# Data is available for how many years?
load_hist.year.unique()

array([2004, 2005, 2006, 2007, 2008], dtype=int64)

The description mentions 4.5 years of hourly data. So, data is not available for the complete 5 years.

In [42]:
# It looks like for 2008, data is available till 31 July - This is incorrect!

import numpy as np
print(np.max(load_hist.loc[load_hist.year == 2008, "month"]), np.max(load_hist.loc[load_hist.year == 2008, "day"]))

7 31


In [5]:
# Monthly Distribution - All Years
for i in [2004, 2005, 2006, 2007, 2008]:
    print(load_hist.loc[load_hist.year == i, "month"].unique())

[ 1  2  3  4  5  6  7  8  9 10 11 12]
[ 1  2  3  4  5  6  7  8  9 10 11 12]
[ 1  2  3  4  5  6  7  8  9 10 11 12]
[ 1  2  3  4  5  6  7  8  9 10 11 12]
[1 2 3 4 5 6 7]


In [41]:
# Latest Available data
load_hist.loc[(load_hist.year==2008) & (load_hist.month==6),]

zone_id  year  month  day      h1      h2      h3      h4      h5  \
1613         1  2008      6    1  15,136  13,315  12,376  11,733  11,427   
1614         1  2008      6    2  12,864  11,868  11,419  11,093  11,236   
1615         1  2008      6    3  13,368  12,074  11,245  10,970  11,310   
1616         1  2008      6    4  13,446  12,469  11,891  11,342  11,733   
1617         1  2008      6    5  15,742  14,623  13,606  12,893  12,919   
...        ...   ...    ...  ...     ...     ...     ...     ...     ...   
32988       20  2008      6   26  78,902  73,387  69,721  68,336  69,365   
32989       20  2008      6   27  89,064  81,762  77,893  75,500  75,549   
32990       20  2008      6   28  84,385  78,279  74,501  72,190  72,121   
32991       20  2008      6   29  79,208  73,644  70,444  68,279  67,600   
32992       20  2008      6   30  76,178  70,839  67,356  66,952  68,682   

           h6  ...      h15      h16      h17      h18      h19      h20  \
1613   11,296  ...   24,317   24,950   24,300   24,455   23,978   22,535   
1614   12,486  ...   21,077   22,324   23,778   24,967   24,971   23,689   
1615   12,502  ...   20,401   21,094   21,476   21,650   21,702   21,241   
1616   12,973  ...   26,868   28,291   29,350   29,697   29,716   28,237   
1617   14,583  ...   30,751   32,340   33,218   34,073   33,878   32,716   
...       ...  ...      ...      ...      ...      ...      ...      ...   
32988  76,007  ...  118,403  123,263  127,387  132,327  141,619  136,113   
32989  80,140  ...  111,177  116,193  122,072  124,814  131,446  125,277   
32990  73,491  ...  125,199  119,400  123,975  130,909  119,595  111,219   
32991  69,436  ...  105,023   98,844  104,201  115,496  113,376  110,741   
32992  74,728  ...      NaN      NaN      NaN      NaN      NaN      NaN   

           h21      h22      h23      h24  
1613    20,719   19,278   17,037   14,707  
1614    22,818   22,187   18,627   15,273  
1615    19,629   18,392   16,907   14,758  
1616    26,739   25,868   21,987   18,207  
1617    30,921   29,841   26,505   22,443  
...        ...      ...      ...      ...  
32988  130,257  128,428  114,889  100,105  
32989  118,750  117,213  107,443   94,293  
32990  106,798  104,711   97,878   87,845  
32991  107,334  106,312   97,012   85,409  
32992      NaN      NaN      NaN      NaN  

[600 rows x 28 columns]

In [43]:
# Data for July is NaN
load_hist.loc[(load_hist.year==2008) & (load_hist.month==7),]

zone_id  year  month  day   h1   h2   h3   h4   h5   h6  ...  h15  h16  \
1643         1  2008      7    1  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
1644         1  2008      7    2  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
1645         1  2008      7    3  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
1646         1  2008      7    4  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
1647         1  2008      7    5  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
...        ...   ...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32995       20  2008      7    3  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
32996       20  2008      7    4  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
32997       20  2008      7    5  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
32998       20  2008      7    6  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
32999       20  2008      7    7  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   

       h17  h18  h19  h20  h21  h22  h23  h24  
1643   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1644   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1645   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1646   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1647   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...    ...  ...  ...  ...  ...  ...  ...  ...  
32995  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
32996  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
32997  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
32998  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
32999  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[140 rows x 28 columns]

In [6]:
# Day Distribution - All Years

num_of_days = []

for i in [2004, 2005, 2006, 2007]:
    for j in [1,2,3,4,5,6,7,8,9,10,11,12]:
        num_of_days.append(len(load_hist.loc[(load_hist.year == i) & (load_hist.month == j), "day"].unique()))

print(np.min(num_of_days), np.max(num_of_days), np.mean(num_of_days))

28 31 30.4375


In [7]:
# 12 months * 4 years
len(num_of_days)

48

In [8]:
# Number of unique zones - Zone 21 is not present in this file
load_hist.zone_id.unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20], dtype=int64)

In [9]:
# Zone IDs are stored as an int type
type(load_hist.loc[0,"zone_id"])

numpy.int64

In [10]:
# Trying to find Zone 21 - Does not exist!
load_hist.loc[load_hist.zone_id == 21,]

Empty DataFrame
Columns: [zone_id, year, month, day, h1, h2, h3, h4, h5, h6, h7, h8, h9, h10, h11, h12, h13, h14, h15, h16, h17, h18, h19, h20, h21, h22, h23, h24]
Index: []

[0 rows x 28 columns]

There is no spatial information available for the zones mentioned here. According to the official description, the local information is deliberately left undisclosed.

The data description indicates the following weeks to be missing:<br>
2005/3/6–2005/3/12, 2005/6/20–2005/6/26, 2005/9/10–2005/9/16, 2005/12/25–2005/12/31, 2006/2/13–2006/2/19, 2006/5/25–2006/5/31, 2006/8/2–2006/8/8, and 2006/11/22–2006/11/28 <br><br>
This can be seen below.

In [26]:
# Checking missing data
load_hist.loc[(load_hist.year==2005) & (load_hist.month==3) & (load_hist.day >= 6) & (load_hist.day <= 12),]

zone_id  year  month  day   h1   h2   h3   h4   h5   h6  ...  h15  h16  \
430          1  2005      3    6  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
431          1  2005      3    7  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
432          1  2005      3    8  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
433          1  2005      3    9  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
434          1  2005      3   10  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
...        ...   ...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
31782       20  2005      3    8  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
31783       20  2005      3    9  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
31784       20  2005      3   10  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
31785       20  2005      3   11  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
31786       20  2005      3   12  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   

       h17  h18  h19  h20  h21  h22  h23  h24  
430    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
431    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
432    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
433    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
434    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...    ...  ...  ...  ...  ...  ...  ...  ...  
31782  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
31783  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
31784  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
31785  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
31786  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[140 rows x 28 columns]

This means that we will need to find a way to deal with the missing values here. <br>
- In any month of any year, the missing values do not exceed a week.
- Only four months in a year seem to show this tendency, rest of the months have full data
- Missing data only exists for selected months of 2005 and 2006

## Holiday List File

In [11]:
# Importing the File - can be used as an explanatory variable

hol_list = pd.read_csv(r"C:\Users\singh\Desktop\TUD (All Semesters)\Courses - Semester 6 (TU Dresden)\Thesis Work\Dataset\GEFCom2012\GEFCOM2012_Data\Load\Holiday_List.csv")
hol_list.head()

Unnamed: 0                 2004  \
0                       New Year's Day  Thursday, January 1   
1  Birthday of Martin Luther King, Jr.   Monday, January 19   
2                Washington's Birthday  Monday, February 16   
3                         Memorial Day       Monday, May 31   
4                     Independence Day       Monday, July 5   

                        2005                 2006                 2007  \
0  Friday, December 31, 2004    Monday, January 2    Monday, January 1   
1         Monday, January 17   Monday, January 16   Monday, January 15   
2        Monday, February 21  Monday, February 20  Monday, February 19   
3             Monday, May 30       Monday, May 29       Monday, May 28   
4             Monday, July 4      Tuesday, July 4    Wednesday, July 4   

                  2008  
0   Tuesday, January 1  
1   Monday, January 21  
2  Monday, February 18  
3       Monday, May 26  
4       Friday, July 4

So the row names are indexed as holiday names, while the columns are marked by year numbers.<br>
This suggests that holidays that repeat annually have been considered here.

In [12]:
# Format of the data saved - the data might require conversion to datetime
print(hol_list.loc[0,"2004"])
type(hol_list.loc[0,"2004"])

Thursday, January 1


str

The datetime information is not directly known, until broken down into individual year, month and day number.

In [13]:
# Total holidays considered
hol_list.shape

(10, 6)

## Temperature History File
Measured in Fahrenheit.

In [14]:
# Importing the file - literature suggests that it' a a strong explanatory variable

temp_hist = pd.read_csv(r"C:\Users\singh\Desktop\TUD (All Semesters)\Courses - Semester 6 (TU Dresden)\Thesis Work\Dataset\GEFCom2012\GEFCOM2012_Data\Load\temperature_history.csv")
temp_hist.head()

station_id  year  month  day  h1  h2  h3  h4  h5  h6  ...   h15   h16  \
0           1  2004      1    1  46  46  45  41  39  35  ...  55.0  55.0   
1           1  2004      1    2  43  44  46  46  47  47  ...  54.0  56.0   
2           1  2004      1    3  45  46  46  44  43  46  ...  69.0  68.0   
3           1  2004      1    4  63  62  62  62  60  60  ...  71.0  72.0   
4           1  2004      1    5  64  63  65  64  64  64  ...  66.0  66.0   

    h17   h18   h19   h20   h21   h22   h23   h24  
0  52.0  46.0  40.0  40.0  39.0  38.0  40.0  41.0  
1  57.0  53.0  50.0  47.0  46.0  45.0  45.0  45.0  
2  68.0  65.0  64.0  63.0  62.0  63.0  63.0  62.0  
3  71.0  68.0  67.0  67.0  65.0  64.0  65.0  64.0  
4  66.0  66.0  63.0  54.0  52.0  49.0  47.0  47.0  

[5 rows x 28 columns]

In [15]:
temp_hist.shape

(18073, 28)

In [16]:
# Seems to resemble the load history df
temp_hist.year.unique()

array([2004, 2005, 2006, 2007, 2008], dtype=int64)

In [17]:
# Here, the latest data is available till 30 June
print(np.max(temp_hist.loc[temp_hist.year == 2008, "month"]),np.max(temp_hist.loc[(temp_hist.year == 2008)&(temp_hist.month == 6), "day"]))

6 30


In [18]:
# Monthly Distribution - All Years
for i in [2004, 2005, 2006, 2007, 2008]:
    print(temp_hist.loc[temp_hist.year == i, "month"].unique())

[ 1  2  3  4  5  6  7  8  9 10 11 12]
[ 1  2  3  4  5  6  7  8  9 10 11 12]
[ 1  2  3  4  5  6  7  8  9 10 11 12]
[ 1  2  3  4  5  6  7  8  9 10 11 12]
[1 2 3 4 5 6]


<b>Note:</b> Data is not included for July 2008.

In [85]:
# For June 2008, are all days included? - Yes!
temp_hist.loc[(temp_hist.year == 2008) & (temp_hist.month == 6), "day"].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30], dtype=int64)

### Important Section
This section tells us where the values get cut-off, and transition to NaNs.

In [89]:
# For 30 June 2008, are all hours of the day included? - No!
temp_hist.loc[(temp_hist.year == 2008) & (temp_hist.month == 6) & (temp_hist.day == 30), ]

station_id  year  month  day  h1  h2  h3  h4  h5  h6  ...  h15  h16  \
1642            1  2008      6   30  74  74  74  74  74  75  ...  NaN  NaN   
3285            2  2008      6   30  66  65  64  64  64  63  ...  NaN  NaN   
4928            3  2008      6   30  77  76  76  75  75  74  ...  NaN  NaN   
6571            4  2008      6   30  72  72  72  72  72  71  ...  NaN  NaN   
8214            5  2008      6   30  73  74  74  73  72  72  ...  NaN  NaN   
9857            6  2008      6   30  70  70  70  70  70  68  ...  NaN  NaN   
11500           7  2008      6   30  74  73  72  71  71  71  ...  NaN  NaN   
13143           8  2008      6   30  72  71  71  71  70  71  ...  NaN  NaN   
14786           9  2008      6   30  75  72  72  69  69  69  ...  NaN  NaN   
16429          10  2008      6   30  70  70  70  70  70  70  ...  NaN  NaN   
18072          11  2008      6   30  70  65  64  64  64  64  ...  NaN  NaN   

       h17  h18  h19  h20  h21  h22  h23  h24  
1642   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3285   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4928   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
6571   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
8214   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
9857   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
11500  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
13143  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
14786  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
16429  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
18072  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[11 rows x 28 columns]

In [96]:
# which hours for 30 June 2008 are not included?
temp_hist.loc[(temp_hist.year == 2008) & (temp_hist.month == 6) & (temp_hist.day == 30) & (temp_hist.station_id == 1), ["h4","h5","h6","h7","h8","h9","h20","h21","h22"]]

h4  h5  h6  h7  h8  h9  h20  h21  h22
1642  74  74  75 NaN NaN NaN  NaN  NaN  NaN

So, values begin to go missing from <b> hour 7 </b> onwards.

In [19]:
# What are the station ids? - Are they same as zone ids?
temp_hist.station_id.unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int64)

No, station IDs are different from the zone IDs. There might be multiple zones present within the perimieter of a station.

What is seen:
- There is no spatial information available for the weather stations mentioned here.
- According to the official description, the local information is deliberately left undisclosed.
- The forecasting horizon begins from 30 June 2008 @7:00 hour, for which the temperature data is made unavailable.

## Load Benchmark File
Predicted hourly loads from 2008/7/1 to 2008/7/7 (First Week of July 2008). The weight column shows the weights assigned to different weeks and levels. <br>
The benchmark file contains backcasted values for the missing values present in the <i>load_history</i> file.

In [30]:
# Importing the File
load_bench = pd.read_csv(r"C:\Users\singh\Desktop\TUD (All Semesters)\Courses - Semester 6 (TU Dresden)\Thesis Work\Dataset\GEFCom2012\GEFCOM2012_Data\Load\Load_benchmark.csv")
load_bench.head()

id  zone_id  year  month  day      h1      h2      h3      h4      h5  ...  \
0   1        1  2005      3    6   20505   19445   19373   18833   19962  ...   
1   2        2  2005      3    6  167016  163275  164618  164731  165840  ...   
2   3        3  2005      3    6  180211  176174  177624  177745  178942  ...   
3   4        4  2005      3    6     530     496     489     460     492  ...   
4   5        5  2005      3    6    8585    8346    8564    8638    8793  ...   

      h15     h16     h17     h18     h19     h20     h21     h22     h23  \
0   14333   13606   14009   15133   16097   16949   16895   16322   14969   
1  145182  140364  141993  148293  151031  158234  162813  158044  148985   
2  156652  151453  153210  160008  162963  170735  175675  170530  160756   
3     497     472     479     512     535     542     552     536     468   
4    6835    6486    6679    7343    7723    8347    8536    7856    6816   

      h24  
0   14016  
1  140408  
2  151501  
3     415  
4    5974  

[5 rows x 29 columns]

In [74]:
# There appears to be no presence of any weight column!
load_bench.columns

Index(['id', 'zone_id', 'year', 'month', 'day', 'h1', 'h2', 'h3', 'h4', 'h5',
       'h6', 'h7', 'h8', 'h9', 'h10', 'h11', 'h12', 'h13', 'h14', 'h15', 'h16',
       'h17', 'h18', 'h19', 'h20', 'h21', 'h22', 'h23', 'h24'],
      dtype='object')

Note: 
- Do not confuse the <u>id</u> column with <i>station ids</i> mentioned in the temperature data file.
- The "id" column begins with 1 instead of 0.


In [31]:
load_bench.shape

(1323, 29)

In [32]:
load_bench.id.unique()

array([   1,    2,    3, ..., 1321, 1322, 1323], dtype=int64)

In [34]:
# 20 zones, with 21st being the aggregated value
load_bench.zone_id.unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21], dtype=int64)

In [58]:
# Value for zone 21: 6 Mar, 2005
load_bench.loc[(load_bench.year==2005)&(load_bench.month==3)&(load_bench.day==6)&(load_bench.zone_id==21), "h1"]

20    1695779
Name: h1, dtype: int64

In [64]:
# Values added over all 20 zones: 6 Mar, 2005
np.sum(load_bench.loc[(load_bench.year==2005)&(load_bench.month==3)&(load_bench.day==6)&(load_bench.zone_id!=21), "h1"])

1695779

Yes, so the values are added over 20 zones for zone 21.

In [44]:
# So, both the missing values as well as the values for the forecast horizon have been predicted
load_bench.year.unique()

array([2005, 2006, 2008], dtype=int64)

In [49]:
# Starting prediction month for 2008 - July

load_bench.loc[load_bench.year == 2008, "month"].unique()

array([7], dtype=int64)

In [67]:
# Hence, forecasts are made only for the first week!

load_bench.loc[load_bench.year == 2008, "day"].unique()

array([1, 2, 3, 4, 5, 6, 7], dtype=int64)

In [71]:
# Predictions for 2008 - 21 zones * 7 days (147 rows)

load_bench.loc[(load_bench.year == 2008), ]

id  zone_id  year  month  day       h1       h2       h3       h4  \
1176  1177        1  2008      7    1    16165    14756    13891    13629   
1177  1178        2  2008      7    1   151110   142943   139773   137593   
1178  1179        3  2008      7    1   163047   154235   150815   148463   
1179  1180        4  2008      7    1      345      322      310      310   
1180  1181        5  2008      7    1     5484     4846     4649     4505   
...    ...      ...   ...    ...  ...      ...      ...      ...      ...   
1318  1319       17  2008      7    7    30944    28246    28037    27780   
1319  1320       18  2008      7    7   185011   173388   158009   155819   
1320  1321       19  2008      7    7    68060    61068    58680    55826   
1321  1322       20  2008      7    7    68701    65929    63832    63483   
1322  1323       21  2008      7    7  1442307  1351579  1280047  1267810   

           h5  ...      h15      h16      h17      h18      h19      h20  \
1176    13788  ...    28089    29973    30321    31031    31890    31333   
1177   140533  ...   231241   233659   237775   239012   236141   231889   
1178   151635  ...   249510   252119   256560   257895   254797   250209   
1179      316  ...      504      520      547      579      582      591   
1180     4765  ...    10668    10992    11747    12303    12740    12429   
...       ...  ...      ...      ...      ...      ...      ...      ...   
1318    27297  ...    43393    42021    43189    45038    46810    45232   
1319   156865  ...   271433   273698   292638   299690   318386   307954   
1320    55533  ...   105985   100756   105238   109228   117191   113689   
1321    65513  ...   108401   107718   109024   114642   119606   116721   
1322  1284920  ...  2109270  2126963  2185141  2210245  2232494  2199402   

          h21      h22      h23      h24  
1176    27613    24839    22432    19215  
1177   224521   210558   191744   171758  
1178   242258   227192   206893   185327  
1179      581      556      489      420  
1180    11739    10475     8865     7160  
...       ...      ...      ...      ...  
1318    44513    42431    38860    34609  
1319   297351   278515   238395   203476  
1320   106420   100171    91445    74506  
1321   114536   107361    94342    82934  
1322  2155061  2021884  1820705  1603514  

[147 rows x 29 columns]

This explains why there are 147 rows for 2008. Now, what about 2005 and 2006?

In [75]:
# Backcasted values for the year 2005

load_bench.loc[(load_bench.year == 2005), ]

id  zone_id  year  month  day       h1       h2       h3       h4  \
0      1        1  2005      3    6    20505    19445    19373    18833   
1      2        2  2005      3    6   167016   163275   164618   164731   
2      3        3  2005      3    6   180211   176174   177624   177745   
3      4        4  2005      3    6      530      496      489      460   
4      5        5  2005      3    6     8585     8346     8564     8638   
..   ...      ...   ...    ...  ...      ...      ...      ...      ...   
583  584       17  2005     12   31    35112    34272    34533    35387   
584  585       18  2005     12   31   222879   215612   213043   225409   
585  586       19  2005     12   31    89619    89682    88620    91190   
586  587       20  2005     12   31    82722    81100    80932    82278   
587  588       21  2005     12   31  1638297  1596759  1580101  1615015   

          h5  ...      h15      h16      h17      h18      h19      h20  \
0      19962  ...    14333    13606    14009    15133    16097    16949   
1     165840  ...   145182   140364   141993   148293   151031   158234   
2     178942  ...   156652   151453   153210   160008   162963   170735   
3        492  ...      497      472      479      512      535      542   
4       8793  ...     6835     6486     6679     7343     7723     8347   
..       ...  ...      ...      ...      ...      ...      ...      ...   
583    35984  ...    28052    27486    28590    31122    33102    33605   
584   236695  ...   191529   189749   198392   227330   243680   245347   
585    88464  ...    66314    68012    70234    78637    85740    87962   
586    82778  ...    84162    83932    87581    92049    93216    92913   
587  1642501  ...  1510424  1494381  1543247  1707082  1764134  1760578   

         h21      h22      h23      h24  
0      16895    16322    14969    14016  
1     162813   158044   148985   140408  
2     175675   170530   160756   151501  
3        552      536      468      415  
4       8536     7856     6816     5974  
..       ...      ...      ...      ...  
583    33133    32287    34029    33014  
584   254833   257494   251683   245106  
585    87447    90314    91392    83589  
586    91152    88060    85399    79110  
587  1755837  1741071  1703950  1592023  

[588 rows x 29 columns]

In 2005, there are 7 days of NaNs for Mar, Jun, Sep and Dec.<br>
So, total missing values should be 7 days * 21 zones * 4 months = 588

## Load Solution File - Still not understood!

- This file contains the actual load data values i.e. actual hourly loads from 2008/7/1 to 2008/7/7.
- This file follows the same format as the <u> load benchmarking file</u>.

Question: What does the <i>weight</i> column actually mean?<br>
Reply: The weight column indicates the weight assigned to different weeks.<br>
Query: As shown towards the end of this section, even within the same period of days, there are different weight. Why?

In [20]:
# Importing
load_sol = pd.read_csv(r"C:\Users\singh\Desktop\TUD (All Semesters)\Courses - Semester 6 (TU Dresden)\Thesis Work\Dataset\GEFCom2012\GEFCOM2012_Data\Load\Load_solution.csv")
load_sol.head()

id  zone_id  year  month  day      h1      h2      h3      h4      h5  ...  \
0   1        1  2005      3    6   19964   19544   19390   19442   19755  ...   
1   2        2  2005      3    6  162096  160890  160924  158962  163197  ...   
2   3        3  2005      3    6  174901  173600  173637  171521  176090  ...   
3   4        4  2005      3    6     528     499     469     486     497  ...   
4   5        5  2005      3    6    9061    8697    8595    8669    8941  ...   

      h16     h17     h18     h19     h20     h21     h22     h23     h24  \
0   13955   13712   14372   16392   18253   18355   17157   16089   15146   
1  148210  149373  153728  171318  175893  175858  166342  155411  145988   
2  159918  161173  165872  184853  189788  189752  179483  167689  157521   
3     450     446     459     547     584     557     564     485     470   
4    6799    6821    7513    9078    9787    9346    8559    7470    6591   

   weight  
0       1  
1       1  
2       1  
3       1  
4       1  

[5 rows x 30 columns]

In [21]:
load_sol.shape

(1323, 30)

In [22]:
load_sol.columns

Index(['id', 'zone_id', 'year', 'month', 'day', 'h1', 'h2', 'h3', 'h4', 'h5',
       'h6', 'h7', 'h8', 'h9', 'h10', 'h11', 'h12', 'h13', 'h14', 'h15', 'h16',
       'h17', 'h18', 'h19', 'h20', 'h21', 'h22', 'h23', 'h24', 'weight'],
      dtype='object')

In [29]:
# 21st zone is the sum of all 21 zones
load_sol.zone_id.unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21], dtype=int64)

In [99]:
# What are the unique weight values?
load_sol.weight.unique()

array([  1,  20,   8, 160], dtype=int64)

In [109]:
# In March 2005, which period of days have a weight of 1?
load_sol.loc[(load_sol.year == 2005) & (load_sol.month == 3) & (load_sol.weight == 1), "day"].unique()

array([ 6,  7,  8,  9, 10, 11, 12], dtype=int64)

In [111]:
# In March 2005, which period of days have a weight of 20?
load_sol.loc[(load_sol.year == 2005) & (load_sol.month == 3) & (load_sol.weight == 20), "day"].unique()

array([ 6,  7,  8,  9, 10, 11, 12], dtype=int64)

This means that the same period has different weights. Something else must also be a factor in deciding the weight. Maybe it is the zone_id!

In [119]:
# Even on a particular day, there are multiple weights
load_sol.loc[(load_sol.year == 2005) & (load_sol.month == 3) & (load_sol.day == 6), "weight"].unique()

array([ 1, 20], dtype=int64)

In [122]:
# The higher weight is given to zone 21!
load_sol.loc[(load_sol.year == 2005) & (load_sol.month == 3) & (load_sol.day == 6) & (load_sol.weight == 20), "zone_id"].unique()

array([21], dtype=int64)

In [126]:
# What do weights look like on a different day of the same month?
load_sol.loc[(load_sol.year == 2005) & (load_sol.month == 3), "day"].unique()

array([ 6,  7,  8,  9, 10, 11, 12], dtype=int64)

## Temperature Solution File
- This file does not include the actual temperature values present in the forecasting horizon.
- It only includes the forecasted values generated using past data.
- The dataset includes the temperature data from 2008/6/30 7:00 to 2008/7/7 24:00.

In [76]:
# Importing the file

temp_sol = pd.read_csv(r"C:\Users\singh\Desktop\TUD (All Semesters)\Courses - Semester 6 (TU Dresden)\Thesis Work\Dataset\GEFCom2012\GEFCOM2012_Data\Load\temperature_solution.csv")
temp_sol.head()

station_id            datetime       date  year  month  day  hour  T0_p1
0           1   30Jun2008:7:00:00  30Jun2008  2008      6   30     7     75
1           1   30Jun2008:8:00:00  30Jun2008  2008      6   30     8     76
2           1   30Jun2008:9:00:00  30Jun2008  2008      6   30     9     78
3           1  30Jun2008:10:00:00  30Jun2008  2008      6   30    10     79
4           1  30Jun2008:11:00:00  30Jun2008  2008      6   30    11     80

In [79]:
# So we have 11 stations with us, as in the temperature history file
temp_sol.station_id.unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int64)

In [81]:
# Time Range?
print(temp_sol.year.unique(), temp_sol.month.unique())

[2008] [6 7]


In [83]:
# For June, which days are included? - Only 30th June
temp_sol.loc[(temp_sol.year == 2008) & (temp_sol.month == 6), "day"].unique()

array([30], dtype=int64)

In [98]:
# The solution values should begin from 30 June 2008, 7:00 hours. - Yes, that's the case here.
temp_sol.loc[(temp_sol.year == 2008) & (temp_sol.month == 6) & (temp_sol.day == 30), "hour"]

0        7
1        8
2        9
3       10
4       11
        ..
1873    20
1874    21
1875    22
1876    23
1877    24
Name: hour, Length: 198, dtype: int64

Either this file can be used for temperature estimates for the forecasting horizon. Alternatively, estimates could be reached using your own approach.